### PR TITLE
Internal: Future-proof Fmt API in case Fmt.t goes abstract

### PR DESCRIPTION
**Pull request by emillon**
Hi,

This is a set of three independent improvements to `Fmt` in case `Fmt.t` becomes abstract in the future, but I think they improve the situation even if that's not the case:

- transform `list_pn` arguments to required options (from optional ones). These arguments are always passed, so optional arguments do not have any advantages. Actually, when `Fmt.t` is made abstract, this confuses the argument erasure check since the last argument becomes `?next`, which can't be erased (before, the last argument was the formatter).
- monomorphize `( $ )` and move it to `Fmt`. It was only used like this in the code base.
- add a `Fmt.set_options` helper. This ensures there is no sequencing with formatters done in `Translation_unit` itself. This iteration is done with `( $ )` instead, making sure it will continue working if `Fmt.t` is abstract.

### CODE REVIEW

File: src/Translation_unit.ml [hunk 1]
```diff
@@ -230,6 +230,17 @@ let check_all_locations fmt cmts_t =
           "Warning: Some locations have not been considered\n%!" ;
         List.iter ~f:print (List.sort l ~compare:Location.compare)
 
+let with_optional_box_debug ~box_debug k fs =
+  if box_debug then Fmt.with_box_debug k fs else k fs
```
Comment (Julow) [inline src/Translation_unit.ml]: You shouldn't take `fs` argument here, `with_box_debug` is `t -> t`.

**Comment by emillon** [comment]
There's more work in other modules to make `Fmt.t` abstract, I'd like to keep this PR small enough.

File: src/Translation_unit.ml (hunk 1 above)
Comment (emillon) [reply src/Translation_unit.ml]: good idea!

File: src/Fmt.ml [hunk 2]
```diff
@@ -257,3 +260,9 @@ let fill_text ?epi text =
                   | _ -> noop)))
       $ fmt_if (Char.is_whitespace text.[String.length text - 1]) " "
       $ opt epi Fn.id )
+
+let set_options ~margin ~max_indent ~prefix fs =
+  set_margin margin fs ;
+  Option.iter max_indent ~f:(fun x -> set_max_indent x fs) ;
```
[the remark this answers is not in the document]
Comment (emillon) [reply src/Fmt.ml]: sure, these functions are already exposed

[the remark this answers is not in the document]
Comment (emillon) [reply src/Fmt.ml]: done, this looks better indeed

[the remark this answers is not in the document]
Comment (emillon) [reply src/Fmt.ml]: hmm actually I'll even inline the call site, the distinction between `set_options` and the rest is a bit too arbitrary

**Comment by emillon** [comment]
Can we keep the original PR title? it explains the reasons of this change, not what the individual changes are. I'll keep the details in the commit message.

**Comment by gpetiot** [comment]
> Can we keep the original PR title? it explains the reasons of this change, not what the individual changes are. I'll keep the details in the commit message.

If the commit message is well detailed sure.

**Pull request merged**